extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java

```diff
@@ -235,25 +241,40 @@ public Uni<Void> nullUni() {
     }
 
     private Uni<Void> persist(ReactiveMongoCollection collection, Object entity) {
+        VersionHandler.handle(entity, getBsonDocument(collection, entity));
```

Comment:
At first, I thought it was a no-op, then I see that VersionHandle.handle creates a version handle wich modify the entity inside it's adjust method called by the constructor.
I find the code path very hard to follow and by looking at the modification inside the MongoOperation class we don't very inderstand what's going on.
Think that many people should open this class without knowledge of how the version mechanism works, this should be more straigthforward to understand.

Reply:
@evanchooly I adjusted the code to reset the version when the update fails, so the user will be able to recovery since there is no changes in the version.

@loicmathieu I really appreciate any tip here about the names/refactoring to make this easier to read, before I named this "of()" since it is a constructor, now it is "handle", but it is not cool for me.

Reply:
I had an idea here, I'm refactoring some things to make this clear :)

Reply:
@loicmathieu what I did was create the `EntityVersionInfo`, and auxiliar class that has info about version, and query to update. The `VersionHandler` produces an `EntityVersionInfo` and handle the version value, increasing or rollback the same in failure situations.

@evanchooly I've changed a lot this code, now every time you have the version annotation, it follows another path, that guarantee the state of the entity for all cases such as persist/persistOrUpdate/update, let me know your considerations.